### PR TITLE
Improve -rbundler/setup performance

### PR DESCRIPTION
**Pull request by segiddins**
@indirect please run your benchmarks against this :)

### CODE REVIEW

File: lib/bundler/definition.rb [hunk 1]
```diff
@@ -581,6 +586,7 @@ def dependencies_for_source_changed?(source, locked_source = source)
     end
 
     def specs_for_source_changed?(source)
+      return false if source.is_a?(Source::Path) && !source.has_spec_files? && Bundler.settings[:frozen]
```
Comment (segiddins) [inline lib/bundler/definition.rb]: I think this was trying to address a different issue, and so it needs test coverage

File: lib/bundler/definition.rb [hunk 2]
```diff
@@ -728,7 +735,7 @@ def converge_locked_specs
         # then we unlock it.
 
         # Path sources have special logic
-        if s.source.instance_of?(Source::Path) || s.source.instance_of?(Source::Gemspec)
+        if (s.source.instance_of?(Source::Path) || s.source.instance_of?(Source::Gemspec)) && (s.source.has_spec_files? || !Bundler.settings[:frozen])
```
Comment (segiddins) [inline lib/bundler/definition.rb]: same thing, different issue & needs tests

File: lib/bundler/rubygems_ext.rb [hunk 3]
```diff
@@ -167,7 +167,11 @@ def matches_spec?(spec)
   class Requirement
     # Backport of performance enhancement added to Rubygems 1.4
     def none?
-      @none ||= (to_s == ">= 0")
+      if @requirements.size == 1
```
Comment (segiddins) [inline lib/bundler/rubygems_ext.rb]: this is taking from the current RubyGems and avoids allocations

**Comment by segiddins** [comment]
Failures appear to be only for Ruby < 2 and not related to the core of these changes

**Comment by segiddins** [comment]
I can't reproduce the failures locally :(

**Comment by segiddins** [comment]
@indirect if we can pinpoint which one of these caused the regression, I'd really like to ship the rest

**Comment by esparta** [comment]
I was able to reproduce one of the failures. And in order to fix it, instead of this:
```diff
diff --git a/spec/runtime/setup_spec.rb b/spec/runtime/setup_spec.rb
index eab66e2..236f600 100644
--- a/spec/runtime/setup_spec.rb
+++ b/spec/runtime/setup_spec.rb
@@ -152,6 +152,7 @@ describe "Bundler.setup" do
         bundler_path.join("gems/bundler-#{Bundler::VERSION}/lib").to_s,
         tmp("rubygems/lib").to_s,
         root.join("../lib").expand_path.to_s,
+        root.join("../lib/bundler/vendor").expand_path.to_s,
       ]
       load_path.map! {|lp| lp.sub(/^#{system_gem_path}/, "") }
```

Should be this:

```diff
diff --git a/spec/runtime/setup_spec.rb b/spec/runtime/setup_spec.rb
index eab66e2..236f600 100644
--- a/spec/runtime/setup_spec.rb
+++ b/spec/runtime/setup_spec.rb
@@ -152,6 +152,7 @@ describe "Bundler.setup" do
         bundler_path.join("gems/bundler-#{Bundler::VERSION}/lib").to_s,
         tmp("rubygems/lib").to_s,
         root.join("../lib").expand_path.to_s,
+        root.join("../bundler/lib/bundler/vendor").expand_path.to_s,
       ]
       load_path.map! {|lp| lp.sub(/^#{system_gem_path}/, "") }
```

The main question is: are we expecting path be on `..lib/bundler/vendor`?  Because the 'fix' seems to easy.

**Comment by indirect** [comment]
@segiddins ping

**Comment by segiddins** [comment]
@indirect we still haven't narrowed down what exactly was causing the performance regression

**Pull request closed without merging**